qiita_db/metadata_template/base_metadata_template.py

```diff
@@ -772,6 +773,119 @@ def _add_common_creation_steps_to_queue(cls, md_template, obj_id,
+
+        Raises
+        ------
+        QiitaDBError
+            If no new samples or new columns are present in `md_template`
+        """
+        # Check if we are adding new samples
+        sample_ids = md_template.index.tolist()
+        curr_samples = set(self.keys())
+        existing_samples = curr_samples.intersection(sample_ids)
+        new_samples = set(sample_ids).difference(existing_samples)
+
+        # Check if we are adding new columns
+        table_name = self._table_name(self._id)
+        # Get the required columns from the DB
+        db_cols = get_table_cols(self._table, conn_handler)
+        # Remove the sample_id and _id_column columns
+        db_cols.remove('sample_id')
+        db_cols.remove(self._id_column)
+        # Get the columns from the dynamic table and do the union with the db
+        curr_cols = set(
+            get_table_cols(table_name, conn_handler)).union(db_cols)
+        headers = md_template.keys().tolist()
+        existing_cols = curr_cols.intersection(headers)
+        new_cols = set(headers).difference(existing_cols)
+
+        if not new_cols and not new_samples:
+            raise QiitaDBError(
+                "No new samples or new columns found in the template. If you "
+                "want to update existing values, you should use the 'update' "
+                "functionality.")
+
+        if new_cols:
+            # If we are adding new columns, add them first (simplifies code)
+            # Sorting the new columns to enforce an order
+            new_cols = sorted(new_cols)
+            datatypes = get_datatypes(md_template.ix[:, new_cols])
+            sql_cols = """INSERT INTO qiita.{0} ({1}, column_name, column_type)
+                          VALUES (%s, %s, %s)""".format(self._column_table,
+                                                        self._id_column)
+            sql_alter = """ALTER TABLE qiita.{0} ADD COLUMN {1} {2}"""
+            for category, dtype in zip(new_cols, datatypes):
+                conn_handler.add_to_queue(
+                    queue_name, sql_cols, (self._id, category, dtype))
+                conn_handler.add_to_queue(
+                    queue_name, sql_alter.format(table_name, category, dtype))
+
+            if existing_samples:
+                warnings.warn(
+                    "No values have been modified for samples '%s'. However, "
+                    "the following columns have been added to them: '%s'"
+                    % (", ".join(existing_samples), ", ".join(new_cols)),
+                    QiitaDBWarning)
+                # The values for the new columns are the only ones that get
+                # added to the database. None of the existing values will be
+                # modified (see update for that functionality)
+                min_md_template = md_template[new_cols].loc[existing_samples]
+                values = as_python_types(min_md_template, new_cols)
+                values.append(existing_samples)
```

Comment:
Can you add an explanation for this and the following line? It's a bit confusing.